### PR TITLE
GH-99205: Mark new interpreters and threads as non-static

### PR DESCRIPTION
**Pull request by brandtbucher**
* Issue: gh-99205

### CODE REVIEW

**Review by kumaraditya303 — APPROVED** [review]
LGTM

File: Python/pystate.c [hunk 1]
```diff
@@ -356,6 +356,7 @@ PyInterpreterState_New(void)
         interp = &runtime->_main_interpreter;
         assert(interp->id == 0);
         assert(interp->next == NULL);
+        assert(interp->_static);
```
Comment (ericsnowcurrently) [inline Python/pystate.c:359]: Good idea.

File: Python/pystate.c [hunk 2]
```diff
@@ -848,6 +851,7 @@ new_threadstate(PyInterpreterState *interp)
         memcpy(tstate,
                &initial._main_interpreter._initial_thread,
                sizeof(*tstate));
+        tstate->_static = false;
```
Comment (ericsnowcurrently) [inline Python/pystate.c:858]: Please add a note pointing to [`PyThreadState_INIT()`](https://github.com/python/cpython/blob/2eee9d9cd7eb1e396fa9a4af7c5fadeeafbdaa38/Include/internal/pycore_runtime_init.h#L68-L73) and indicating that fields set there should be adjusted here as appropriate.

File: Python/pystate.c [hunk 3]
```diff
@@ -370,6 +371,7 @@ PyInterpreterState_New(void)
         // Set to _PyInterpreterState_INIT.
         memcpy(interp, &initial._main_interpreter,
                sizeof(*interp));
+        interp->_static = false;
```
Comment (ericsnowcurrently) [inline Python/pystate.c:376]: Likewise, a note pointing to [`PyInterpreterState_INIT()`](https://github.com/python/cpython/blob/2eee9d9cd7eb1e396fa9a4af7c5fadeeafbdaa38/Include/internal/pycore_runtime_init.h#L48-L66).

**Review by ericsnowcurrently — APPROVED** [review]
LGTM

I've pointed out two places where a comment would be appropriate.  I'm approving the PR under the assumption you'll take care of those.

**Comment by ericsnowcurrently** [comment]
Thanks for fixing this, BTW.

**Comment by brandtbucher** [comment]
@pablogsal, leak tests didn't catch this because we use the "raw" domain to allocate `PyThreadState` and `PyInterpreterState`.

**Comment by pablogsal** [comment]
Oh interesting, can you confirm that changing the domain to memory makes the test fail?

**Comment by brandtbucher** [comment]
> Oh interesting, can you confirm that changing the domain to memory makes the test fail?

You can't. :)

```
Fatal Python error: _PyMem_DebugCalloc: Python memory allocator called without holding the GIL
Python runtime state: preinitialized
```

**Pull request merged**